tests/test_grad.py

```diff
@@ -0,0 +1,196 @@
+    """
+
+    def decorate(fn):
+        def test(args):
+            if not isinstance(args, tuple):
+                args = (args,)
+            # TODO: avoid re-parsing and re-gradding every time
+            graph = parse(fn)
+            gfn = compile(grad(graph))
+            py_result = fn(*map(copy, args))
+            myia_result, bprop_fn = gfn(*map(copy, args))
+            assert py_result == myia_result
+
+            argnames = [p.debug.name for p in graph.parameters]
+
+            test = GradTester(fn=fn,
+                              gfn=bprop_fn,
+                              args=args,
+                              argnames=argnames,
+                              outnames=[fn.__name__])
+            results = test.compare()
+
+            for arg, d in results.items():
+                if not d['match']:
+                    print(f'Argument {arg}:')
+                    print(f'\tFinite differences: {d["difference"]}')
+                    print(f'\tGradient output:    {d["exact"]}')
+                    fail(f'Mismatch in gradients for {arg} (see stdout)')
+
+        m = mark.parametrize('args', list(tests))(test)
+        m.__orig__ = fn
+        return m
+    return decorate
+
+
+@grad_test((13, 14))
+def test_null(x, y):
+    """Test null gradient."""
+    return 10 + 28 / 43
+
+
+@grad_test((3, 4))
+def test_tuple(x, y):
+    """Test multiple outputs."""
+    return (x + y, x - y, x * y, x / y)
+
+
+@grad_test((3, 4, 5))
+def test_expression(x, y, z):
+    """Test a more complex expression."""
+    return x * y + y / z
+
+
+@grad_test((3, 4, 5))
+def test_variables(x, y, z):
+    """Test an expression with temporary variables."""
+    a = x * y
+    b = y * a
+    c = a + b
+    return c / z
+
+
+@grad_test((3, 4, 5))
+def test_shadowing(x, y, z):
+    """Test an expression where variables are shadowed."""
+    x = x * y
+    y = y * z
+    z = x * z
+    return x + y + z
+
+
+@grad_test((3,))
+def test_constant(x):
+    """Test the use of a literal in the expression."""
+    return 18 * x
+
+
+@grad_test((3,))
+def test_dup_args_in_call(x):
+    """The naive gradient update rule fails when a function's arguments
+    contain the same variable more than once."""
+    return x * x
+
+
+@grad_test((3,))
+def test_quadruple_args_in_call(x):
+    """Test that duplicated arguments still cause no problem even if
+    there are four of them."""
+    def g(a, b, c, d):
+        return a * b * c * d
+    return g(x, x, x, x)
+
+
+@grad_test((4, 5))
+def test_simple_closure(a, b):
+    """Test some trivial closures."""
+    def f():
+        return a + 1
+
+    def g():
+        return b + 2
+    return f() * g()
+
+
+@grad_test((4,))
+def test_closure(a):
+    """This is the closure test in the paper."""
+    def x1(b):
+
+        def x4(c):
+            return b
+        return x4
+    x2 = x1(a)
+    x3 = x2(1)
+    return x3
+
+
+# TODO: test when a == b (finite diff won't return anything sensible)
```

Comment:
We could have an option for finite diff to only test the left slope or the right slope, and pass the one that will return the right thing.

Reply:
Right. I think the best option, which Bart mentioned he used for Tangent, would be to compare with autograd. It was simpler for me to just use the finite diff code I already had, but ultimately I think we should compare with autograd or tangent.